driver-core/src/main/com/mongodb/internal/connection/SingleServerCluster.java

```diff
@@ -82,19 +82,21 @@ public void close() {
     private class DefaultServerStateListener implements ServerListener {
         @Override
         public void serverDescriptionChanged(final ServerDescriptionChangedEvent event) {
-            ServerDescription descriptionToPublish = event.getNewDescription();
-            if (event.getNewDescription().isOk()) {
+            ServerDescription newDescription = event.getNewDescription();
+            if (newDescription.isOk()) {
                 if (getSettings().getRequiredClusterType() != ClusterType.UNKNOWN
-                        && getSettings().getRequiredClusterType() != event.getNewDescription().getClusterType()) {
-                    descriptionToPublish = null;
+                        && getSettings().getRequiredClusterType() != newDescription.getClusterType()) {
+                    newDescription = null;
                 } else if (getSettings().getRequiredClusterType() == ClusterType.REPLICA_SET
                         && getSettings().getRequiredReplicaSetName() != null) {
-                    if (!getSettings().getRequiredReplicaSetName().equals(event.getNewDescription().getSetName())) {
-                        descriptionToPublish = null;
+                    if (!getSettings().getRequiredReplicaSetName().equals(newDescription.getSetName())) {
+                        newDescription = ServerDescription.builder(newDescription).type(ServerType.UNKNOWN).setName(null).build();
```

Comment:
The server selector implementations, when the connection is direct, will select any server where isOk == true (see com.mongodb.internal.selector.ReadPreferenceServerSelector#select, for example).  So I think we better set ok to false here as well.

I am, though, troubled by this change, since the server description in the cluster description will almost match, but won't quite match, the one in the DefaultServer.  Not sure what to do about it though.  Perhaps to make it clear in the server selection timeout exception, we should also add an exception to the server description via `com.mongodb.connection.ServerDescription.Builder#exception`.  Something like:

    builder(newDescription).exception(new MongoConfigurationException("Replica set name does not match required replica set name of ...)

Reply:
Updated